tools/fix_mailmap.py

```diff
@@ -0,0 +1,573 @@
+                )
+    return sorted(new_mailmap_lines, key=str.lower)
+
+
+def find_canonical_name(dup_authors: set, canonical_names: set):
+    """Search for a canonical name for a set of authors."""
+    dup_names = tuple({d[1] for d in dup_authors})
+    dup_names_canonical = tuple(d in canonical_names for d in dup_names)
+    canonical_name = None
+    if len(dup_names) == 1:
+        canonical_name = dup_names[0]
+        msg = f"    Only one unique name '{canonical_name}'\n"
+    elif all(dup_names_canonical):
+        names_list = ", ".join(f"'{n}'" for n in dup_names)
+        msg = f"    All {len(names_list)} names are in mailmap: {names_list}\n"
+        msg += "    This tool doesn't handle when normalization merges two unique authors."
+        # If this comes up, either split the list of duplicates,
+        # or do less-aggressive normalization
+        return msg, None
+    elif dup_names_canonical.count(True) == 1:
+        canonical_name = dup_names[dup_names_canonical.index(True)]
+        names_list = ", ".join(f"'{n}'" for n in dup_names if n != canonical_name)
+        msg = f"    One name is in the mailmap: '{canonical_name}'\n"
+        msg += f"        Setting the others ({names_list}) to match it\n"
+    elif dup_names_canonical.count(True) == 0:
+        counted_names = [(d, sum(a[0] for a in dup_authors if a[1] == d)) for d in dup_names]
+        counted_names.sort(key=lambda x: x[1], reverse=True)
+        canonical_name = counted_names[0][0]
+        msg = f"    No names are in mailmap. Choosing '{canonical_name}' as the most common.\n"
+        namecount = (f"'{c[0]}'({c[1]})" for c in counted_names)
+        namecountjoin = ", ".join(namecount)
+        msg += f"        {namecountjoin}\n"
+    else:
+        msg = "     No obvious choice for canonical name. Please resolve manually."
+    return canonical_name, msg
+
+
+def find_canonical_email(dup_authors: set, canonical_emails: set):
+    """Search for a canonical email for a set of authors.
+
+    Prefer:
+        Unique input email
+        Unique email mentioned in mailmap
+        Unique github-noreply email
+        Unique @ibm.com email
+        Most common email from git log
+    """
+    dup_emails = tuple({d[2] for d in dup_authors})
+    dup_emails_canonical = tuple(d in canonical_emails for d in dup_emails)
+    canonical_email = None
+    if len(dup_emails) == 1:
+        canonical_email = dup_emails[0]
+        msg = f"    One unique email <{canonical_email}>\n"
+    elif all(dup_emails_canonical):
+        emails_list = ", ".join(f"<{e}>" for e in dup_emails)
+        msg = f"    All {len(emails_list)} names are in mailmap: {emails_list}\n"
+        msg += "    This tool doesn't handle when normalization merges two unique authors."
+        # If this comes up, either split the list of duplicates,
+        # or do less-aggressive normalization
+    elif dup_emails_canonical.count(True) == 1:
```

Comment:
Do we need to be concerned about performance at all here?  We do a lot of iterating all the way through the tuple.  I'm guessing it probably shouldn't really matter.

Reply:
I dont think we care. It's fast enough since it is only running this on the cases where there is a duplication.